// argos-cli/lib/config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _convict = require('convict');

var _convict2 = _interopRequireDefault(_convict);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const config = (0, _convict2.default)({
  endpoint: {
    doc: 'Argos API endpoint',
    format: 'url',
    default: 'https://api.argos-ci.com',
    env: 'ARGOS_API_ENDPOINT'
  },
  commit: {
    doc: 'Git commit',
    format: String,
    default: '',
    env: 'ARGOS_COMMIT'
  },
  branch: {
    doc: 'Git branch',
    format: String,
    default: '',
    env: 'ARGOS_BRANCH'
  },
  token: {
    doc: 'Repository token',
    format: String,
    default: '',
    env: 'ARGOS_TOKEN'
  }
});

const NODE_ENV = process.env.NODE_ENV || 'development';

if (NODE_ENV !== 'production') {
  config.loadFile(_path2.default.join(__dirname, `../config/${NODE_ENV}.json`));
}
config.validate();

exports.default = config;
